### PR TITLE
[File Upload] server side check for unique names and remove xlsx support

### LINKED ISSUE #13152: Add a server side check for unique names
Validation for unique names should be done in the server

_Originally posted by @tallulahkay in https://github.com/civiform/civiform/pull/13134#discussion_r3119294669_

### PR DESCRIPTION
**Pull request by tallulahkay**
### Description

- Remove xlsx support because Seattle said they no longer need it
- Move unique name generation to the server

### Checklist

Remove any sections below that do not apply to your PR. For sections that do apply, complete all of the appropriate checklist items and check them off. If a checklist item doesn't apply to your PR, leave it unchecked but do not delete it.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label (see [docs](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-appropriate-labels) for more info): < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [ ] Added an additional reviewer from `civiform/eng-admin` as FYI (if this PR includes functionality changes and neither you nor the primary reviewer is an admin)
- [ ] Added an additional reviewer as required if changes potentially affect the security of the application.
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).
- [ ] Ensured PII wasn't added to any new logs, unless it was guarded by `isDevOrStaging`
- [ ] Noted in the PR description which, if any, code in this PR was generated by AI.

### Issue(s) this completes

Fixes #13152

### CODE REVIEW

File: server/app/services/applicant/question/FileUploadQuestion.java [hunk 1]
```diff
@@ -277,6 +279,35 @@ public static String getFileName(String fileKey) {
...
+  private static final Pattern FILE_NAME_DIGIT_SUFFIX_REGEX =
+      Pattern.compile("(.*?)(-\\d+)(\\.[^.]+)?$");
+  // Matches a filename with an extension.
+  private static final Pattern FILE_NAME_REGEX = Pattern.compile("(.*)(\\.[^.]+)$");
+
+  /**
+   * Returns a name derived from {@code name} that does not appear in {@code existingNames},
+   * appending or incrementing a "-N" numeric suffix as needed.
+   */
+  public static String getUniqueName(String name, ImmutableList<String> existingNames) {
```
Comment (gwendolyngoetz) [inline server/app/services/applicant/question/FileUploadQuestion.java:295]: (blocking) validate name is not null or empty, validate existingNames is not null)

File: server/app/controllers/applicant/FileUploadController.java [hunk 2]
```diff
@@ -146,8 +147,13 @@ public CompletionStage<Result> hxSelectFileForUpload(
...
               }
 
+              ImmutableList<String> existingFileNames =
+                  fileUploadQuestion.getOriginalFileNameListValue().orElse(ImmutableList.of());
+              String deduplicatedFileName =
+                  FileUploadQuestion.getUniqueName(originalFileName, existingFileNames);
+
               ImmutableMap<String, String> formData =
-                  fileUploadQuestion.buildFormDataForAdd(fileKey, originalFileName);
+                  fileUploadQuestion.buildFormDataForAdd(fileKey, deduplicatedFileName);
```
Comment (gwendolyngoetz) [inline server/app/controllers/applicant/FileUploadController.java:156]: (question,  blocking) This stores one file original file name on the applicant record and a different one on the file record. Is there a reason they aren't the same? This feels like a bug or setting us up for one down the road.

File: server/app/services/applicant/question/FileUploadQuestion.java [hunk 3]
```diff
@@ -277,6 +279,35 @@ public static String getFileName(String fileKey) {
...
+    while (existingNames.contains(name)) {
+      Matcher digitSuffixMatcher = FILE_NAME_DIGIT_SUFFIX_REGEX.matcher(name);
+      if (digitSuffixMatcher.matches()) {
+        int nextNumber = Integer.parseInt(digitSuffixMatcher.group(2).substring(1)) + 1;
```
Comment (gwendolyngoetz) [inline server/app/services/applicant/question/FileUploadQuestion.java]: (blocking) A valid file with the right name would cause an integer overflow here.

File: server/app/services/applicant/question/FileUploadQuestion.java [hunk 4]
```diff
@@ -277,6 +279,35 @@ public static String getFileName(String fileKey) {
...
+      Pattern.compile("(.*?)(-\\d+)(\\.[^.]+)?$");
+  // Matches a filename with an extension.
+  private static final Pattern FILE_NAME_REGEX = Pattern.compile("(.*)(\\.[^.]+)$");
+
+  /**
+   * Returns a name derived from {@code name} that does not appear in {@code existingNames},
+   * appending or incrementing a "-N" numeric suffix as needed.
+   */
+  public static String getUniqueName(String name, ImmutableList<String> existingNames) {
+    while (existingNames.contains(name)) {
```
Comment (gwendolyngoetz) [inline server/app/services/applicant/question/FileUploadQuestion.java:314]: (thinking out load, maybe blocking) There is potential that we are given a very long file name that is at the upper limit for a valid file name and when we add to it could potentially make the length exceed what the operating system allows. Some browsers may be fine or handle it... :thinking:

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/services/applicant/question/FileUploadQuestion.java:295]: (question, blocking) Do we just want to error if there is no file extension or is a dotfile (it just looks like a file extension like .gitignore)? I'm assuming Tika and that related code is going to fail without one anyway. But if it may simplify things in the former case. The latter case is broken (.gitignore that already exists I think becomes `-2.gitignore`.

File: server/test/services/applicant/question/FileUploadQuestionTest.java [hunk 5]
```diff
@@ -278,4 +278,18 @@ public void getOriginalFileNameForFileKey_noOriginalName_returnsEmpty() {
...
+    assertThat(FileUploadQuestion.getUniqueName("file.txt", ImmutableList.of()))
+        .isEqualTo("file.txt");
+    assertThat(FileUploadQuestion.getUniqueName("file.txt", ImmutableList.of("file.txt")))
+        .isEqualTo("file-2.txt");
+    assertThat(
+            FileUploadQuestion.getUniqueName(
+                "file.txt", ImmutableList.of("file.txt", "file-2.txt")))
+        .isEqualTo("file-3.txt");
+    assertThat(FileUploadQuestion.getUniqueName("file", ImmutableList.of("file")))
+        .isEqualTo("file-2");
```
Comment (gwendolyngoetz) [inline server/test/services/applicant/question/FileUploadQuestionTest.java:314]: One the notes on the `getUniqueName` are handled it would be good to add some more possible file name formats either here or as other test methods to cover some of those odder edge cases.

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 1 above)
Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java:295]: This unique name function is for user display only.  Re: both of these comments, this getUniqueName function runs _after_ the parser, so we would have already failed on null names or no file extension or a dotfile. Adding it here would only be defensive in case this method was ever called in another context. I'm still happy to add it though!

File: server/app/controllers/applicant/FileUploadController.java (hunk 2 above)
Comment (tallulahkay) [reply server/app/controllers/applicant/FileUploadController.java:156]: Well, there's sorta a reason but it's unsatisfying. The reason they aren't the same is because the one stored on the file record just isn't used at all. The source of truth for the display names is on the applicant record and the info used on the file record is the fileKey and acl. 

applicant record stores file key and file display names and file record stores file key and acl

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 3 above)
Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java]: omg like if I uploaded file-2147483647.pdf?? I had to google integer overflow thanks for flagging this

Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java]: do you think we have to worry about someone doing this? The impact is them having an extra weird file name displayed, but someone would have to either upload 2,147,483,647 of the same named file or intentionally use that suffix to mess with us

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 4 above)
Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java:314]: oh interesting. Does it change this thought if this is only used for display? Nothing about this generated filename is stored anywhere besides our database

File: server/app/controllers/applicant/FileUploadController.java (hunk 2 above)
Comment (gwendolyngoetz) [reply server/app/controllers/applicant/FileUploadController.java:156]: Yeah that is unsatisfying. Used or not it's just a small change a few lines down to keep them in sync right? I think we should do that.

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/services/applicant/question/FileUploadQuestion.java:295]: We should do the basic validation from the first comment as a minimum. It would be good to make sure we handle all the cases explicitly either by blocking them at the start or making sure they get numbered as expected right now just in the order of things gets changed down the road.

Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java:295]: I already included the basic validation you suggested but happy to add the other cases

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 3 above)
Comment (gwendolyngoetz) [reply server/app/services/applicant/question/FileUploadQuestion.java]: I don't think it's necessarily an invalid file or malicious. They should have downloaded it somewhere that appended the value. There's still a max cap of files allowed on a file upload (I want to say 50 per question type?). If a user did have `Integer.MAX_VALUE` number of files then this entire function will need to be rewritten. At that size this is going to start running very slow.

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 4 above)
Comment (gwendolyngoetz) [reply server/app/services/applicant/question/FileUploadQuestion.java:314]: Yeah, I think we'll probably be fine in practice

File: server/app/services/applicant/question/FileUploadQuestion.java (hunk 3 above)
Comment (tallulahkay) [reply server/app/services/applicant/question/FileUploadQuestion.java]: I don't think there is a max cap of files allowed on file upload

File: server/test/services/applicant/question/FileUploadQuestionTest.java (hunk 5 above)
Comment (tallulahkay) [reply server/test/services/applicant/question/FileUploadQuestionTest.java:314]: after brainstorming some examples of valid files (I added tests for them) I realized the paradigm of checking if there is already an extension and then incrementing it doesn't work at all, because what if the original file uploaded validly has -345352 or any other number but it's the first file. so I totally changed the regex matching to instead split into a base and extension for every file and check if it already exists.

**Review by gwendolyngoetz — APPROVED** [review]
